presto-main/src/test/java/com/facebook/presto/operator/scalar/FunctionAssertions.java

```diff
@@ -268,6 +277,114 @@ private Object selectUniqueValue(String projection, Type expectedType, Session s
+                    .mapToLong(this::getRetainedSizeOfCachedInstance)
+                    .sum();
+            if (retainedSize > maxRetainedSize) {
+                maxRetainedSize = retainedSize;
+                maxIterationCount = iterationCount;
+            }
+
+            if (maxRetainedSize >= 1048576) {
+                fail(format("The retained size of cached instance of function invocation is likely unbounded: %s", projection));
+            }
+        }
+    }
+
+    private long getRetainedSizeOfCachedInstance(PageProjection projection)
+    {
+        Field[] fields = projection.getClass().getDeclaredFields();
+        long retainedSize = 0;
+        for (Field field : fields) {
+            field.setAccessible(true);
+            String fieldName = field.getName();
+            if (!fieldName.startsWith("__cachedInstance")) {
+                continue;
+            }
+            try {
+                retainedSize += getRetainedSizeOf(field.get(projection));
+            }
+            catch (IllegalAccessException e) {
+                throw new RuntimeException(e);
+            }
+        }
+        return retainedSize;
+    }
+
+    private long getRetainedSizeOf(Object object)
+    {
+        Class type = object.getClass();
+        if (type == PageBuilder.class) {
+            return ((PageBuilder) object).getRetainedSizeInBytes();
+        }
+        else if (type.isArray()) {
+            if (type == int[].class) {
+                return sizeOf((int[]) object);
+            }
+            else if (type == boolean[].class) {
+                return sizeOf((boolean[]) object);
+            }
+            else if (type == byte[].class) {
+                return sizeOf((byte[]) object);
+            }
+            else if (type == long[].class) {
+                return sizeOf((long[]) object);
+            }
+            else if (type == short[].class) {
+                return sizeOf((short[]) object);
+            }
+            else if (type == Block[].class) {
+                Object[] objects = (Object[]) object;
+                return Arrays.stream(objects)
+                        .mapToLong(this::getRetainedSizeOf)
+                        .sum();
+            }
+            else {
+                throw new UnsupportedOperationException(format("Unknown type encountered: %s", type));
+            }
+        }
+        else {
+            long retainedSize = ClassLayout.parseClass(type).instanceSize();
+            Field[] fields = type.getDeclaredFields();
+            for (Field field : fields) {
+                try {
+                    if (field.getType().isPrimitive()) {
+                        continue;
+                    }
+                    field.setAccessible(true);
+                    retainedSize += getRetainedSizeOf(field.get(object));
+                }
+                catch (IllegalAccessException | IllegalArgumentException t) {
```

Comment:
Which call throws `IllegalArgumentException` (unless we make the above `UnsupportedOperationException` an `IllegalArgumentException`)?